Mathlib/Algebra/Algebra/Basic.lean

```diff
@@ -0,0 +1,921 @@
+    mpr := fun H =>
+      H.symm ▸ by
+        apply_fun ⇑h.unit.val using ((Module.End_isUnit_iff _).mp h).injective
+        erw [End_isUnit_apply_inv_apply_of_isUnit]
+        rfl }
+#align module.End_algebra_map_is_unit_inv_apply_eq_iff Module.End_algebraMap_isUnit_inv_apply_eq_iff
+
+theorem End_algebraMap_isUnit_inv_apply_eq_iff' {x : R}
+    (h : IsUnit (algebraMap R (Module.End R M) x)) (m m' : M) :
+    m' = (↑h.unit⁻¹ : Module.End R M) m ↔ m = x • m' :=
+  { mp := fun H => ((congr_arg h.unit H).trans (End_isUnit_apply_inv_apply_of_isUnit h _)).symm
+    mpr := fun H =>
+      H.symm ▸ by
+        apply_fun (↑h.unit : M → M) using ((Module.End_isUnit_iff _).mp h).injective
+        erw [End_isUnit_apply_inv_apply_of_isUnit]
+        rfl }
+#align module.End_algebra_map_is_unit_inv_apply_eq_iff' Module.End_algebraMap_isUnit_inv_apply_eq_iff'
+
+end
+
+end Module
+
+namespace LinearMap
+
+variable {R : Type _} {A : Type _} {B : Type _} [CommSemiring R] [Semiring A] [Semiring B]
+  [Algebra R A] [Algebra R B]
+
+/-- An alternate statement of `linear_map.map_smul` for when `algebra_map` is more convenient to
+work with than `•`. -/
+theorem map_algebraMap_mul (f : A →ₗ[R] B) (a : A) (r : R) :
+    f (algebraMap R A r * a) = algebraMap R B r * f a := by
+  rw [← Algebra.smul_def, ← Algebra.smul_def, map_smul]
+#align linear_map.map_algebra_map_mul LinearMap.map_algebraMap_mul
+
+theorem map_mul_algebraMap (f : A →ₗ[R] B) (a : A) (r : R) :
+    f (a * algebraMap R A r) = f a * algebraMap R B r := by
+  rw [← Algebra.commutes, ← Algebra.commutes, map_algebraMap_mul]
+#align linear_map.map_mul_algebra_map LinearMap.map_mul_algebraMap
+
+end LinearMap
+
+@[simp]
+theorem Rat.smul_one_eq_coe {A : Type _} [DivisionRing A] [ha : Algebra ℚ A] (m : ℚ) :
+    @SMul.smul _ _ Algebra.toSMul m (1 : A) = ↑m :=
+  (@Algebra.smul_def' (self := ha) m 1).trans <| by rw [mul_one]; rfl
+#align rat.smul_one_eq_coe Rat.smul_one_eq_coe
+
+section Nat
+
+variable {R : Type _} [Semiring R]
+
+-- Lower the priority so that `algebra.id` is picked most of the time when working with
+-- `ℕ`-algebras. This is only an issue since `algebra.id` and `algebra_nat` are not yet defeq.
+-- TODO: fix this by adding an `of_nat` field to semirings.
+/-- Semiring ⥤ ℕ-Alg -/
+instance (priority := 99) algebraNat : Algebra ℕ R where
+  commutes' := Nat.cast_commute
+  smul_def' _ _ := nsmul_eq_mul _ _
+  toRingHom := Nat.castRingHom R
+#align algebra_nat algebraNat
+
+instance nat_algebra_subsingleton : Subsingleton (Algebra ℕ R) :=
+  ⟨fun P Q => by ext; simp⟩
+#align nat_algebra_subsingleton nat_algebra_subsingleton
+
+end Nat
+
+namespace RingHom
+
+variable {R S : Type _}
+
+-- porting note: changed `[Ring R] [Ring S]` to `[Semiring R] [Semiring S]`
+-- otherwise, Lean failed to find a `Subsingleton (ℚ →+* S)` instance
+@[simp]
+theorem map_rat_algebraMap [Semiring R] [Semiring S] [Algebra ℚ R] [Algebra ℚ S] (f : R →+* S)
```

Comment:
This sounds like leanprover/lean4#2074, for which our usual workaround is to locally disable the instance causing the bad path. This generalization doesn't make sense mathematically, so I think it would be better to reuse the usual hack.

Reply:
Can you try to do it? There are 2 more failures in this file caused by `[Ring]`. It would be nice to have them fixed (because this would unblock many files) but I failed to do it.

Reply:
It turns out that providing a few implicit args using `@` works.